content/posts/event-driven-scaling-fission-function-using-keda.md

```diff
@@ -0,0 +1,239 @@
+---
+title: "Event Driven Scaling Fission Function Using KEDA"
+date: 2020-07-14T18:09:02+05:30
+author: "Fission"
+categories: ["Fission", "Serverless", "go", "Keda", "Event-driven", "Apache Kafka", "Tutorial", "Faas"]
+description: "Enabling event driven autoscaling with KEDA integration"
+type: "post"
+---
+
+# Introduction
+
+Events and integrations with event sources such as message queues are an important part of running functions. Fission had the MQ integration for invoking functions and this integration was available for Kafka, NATS, and Azure Queue Storage. This integration had a few limitations:
+
+- For every new integration you want to enable - there was one pod running for enabling the integration
+- Autoscaling of the trigger handler was not available.
+- Lastly - one Fission installation could connect to and handle only one instance of a MQ. So for example you could only connect to one Kafka instance in a given Fission installation. 
+
+When we decided to build out these features we came across Keda project and it solved most of the limitations which Fission had in the event integration area. Let’s dive into how it works!
+
+# Architecture
+
+Let’s first understand how the whole thing works. We won’t explain Keda in detail but you can take a look at Keda documentation for details.
+
+
+![architecture](../../../images/event-driven-scaling-fission-function-using-keda/architecture.png)
+
+
+1. The user creates a trigger - for Keda based integration you have to specify the “mqtkind=keda” and add all relevant parameters. These parameters are different for each message queue and hence are encapsulated in a metadata field and follow a key-value format. As soon as you create the  MQ Trigger, Fission creates a ScaledObject and a consumer deployment object which is referenced by ScaledObject. The ScaledObject is a Keda’s way of encapsulating the consumer deployment and all relevant information for connecting to an event source! Keda goes ahead and creates a HPA for the deployment and scales down the deployment to zero.
+2. As the message arrives in the event source - the Keda will scale the HPA and deployment from 0 - to 1 for consuming messages. As more messages arrive the deployment is scaled beyond 1 automatically too.
+3. The deployment is like an connector which consumes messages from the source and then calls a function. 
+4. The function consumes the message and puts the response in response topic and errors in error topic as may be applicable.
```

Comment:
I think you remember saying that connector puts message back in response queue - do you want to change this to reflect that?

Reply:
Added